wsimod/extensions/extensions.py

```diff
@@ -0,0 +1,38 @@
+"""Extensions module for decorators and subclasses.
+
+Example use for decorators:
+>>> from wsimod.extensions import extensions as extend
+>>> @extend.node_attribute(obj=my_fwtw, attribute_name="pull_distributed")
+>>> def new_distributed(pull_distributed, vqip):
+>>>      return pull_distributed(vqip, tag="FWTW")
+"""
+
+
+def node_attribute(obj, attribute_name: str):
```

Comment:
This function should accept a string indicating the node name to update, not a fully formed object. Of course, there has to be a way for the function to get the right object. I would suggest to create a singleton class - a sort of global variable, but a bit tidier -, from where this one pulls the information it needs. A minimal implementation would be something like:

```python
class ExtensionsHandler:
    __slots__ = "_model"
    _instance: Optional["ExtensionsHandler"] = None

    def __new__(cls, model: Optional[Model] = None):
        if not cls._instance:
            if not model:
                raise ValueError("The extension handler needs to be initialised with a model.")
            cls._instance = super().__new__(cls)
            cls._instance._model = model
        return cls._instance

    def __init__(self, model: Optional[Model] = None):
        # Do not assign `model` here.
        self._model: Model

    def get_node(self, name: str) -> Node:
        return self._model.nodes[name]
```

Now, a few more things:

In the `__init__` method of `Model` class we initialise the handler as the very first thing to do. You do not need to assign it to anything, as it is just a question of making the extension handler aware of what model is to be extended:

```python 
class Model(...):
    def __init__(self, ...):
         ExtensionsHandler(self)
```

And then, within the decorators, now taking a string as input, we do:

```python
def node_attribute(name: str, attribute_name: str):
    def decorator(...):
        obj = ExtensionsHandler().get_node(name)
        ...
```

You can implement a more complex logic, add handlers for arcs, etc, but this would be the basic idea. In summary:
- You register your model with the extension handler.
- You use the extension handler in the decorators to pull the relevant node/arc other from the model you are running.

Does any of this makes sense?

Reply:
As I explained above, if you take as input instantiated objects, you obviously need to already have access to the formed objects in order to use the decorator. That is the case in the tutorial since you are creating all the components programmatically, so you can pass the object directly, but it will not be true when including the extensions in an extension file. Loading that file in order to apply the extensions will fail as the instantiated object is not available _in the module_ when loading it. And it needs to be if it is an input to the decorator itself.

Just try to put your decorated function using instantiated objects in a separate file and load that file as a module. You will get something like `NameError: name "my_fwtw" is not defined`.